lib/hooks/userspace/hook-grpc.js

```diff
@@ -129,6 +129,298 @@ function makeClientConstructorWrap(makeClientConstructor) {
   };
 }
 
+// # Server
+
+/**
+ * A helper function to record metadata in a trace span. The return value of this
+ * function can be used as the 'wrapper' argument to wrap sendMetadata.
+ * sendMetadata is a member of each of ServerUnaryCall, ServerWriteableStream,
+ * ServerReadableStream, and ServerDuplexStream.
+ * @param rootContext The span object to which the metadata should be added.
+ * @returns {Function} A function that returns a wrapped form of sendMetadata.
+ */
+function sendMetadataWrapper(rootContext) {
+  return function (sendMetadata) {
+    return function sendMetadataTrace(responseMetadata) {
+      if (rootContext) {
+        rootContext.addLabel('metadata', JSON.stringify(responseMetadata.getMap()));
+      } else {
+        agent.logger.info('gRPC: No root context found in sendMetadata');
+      }
+      return sendMetadata.apply(this, arguments);
+    };
+  };
+}
+
+/**
+ * Wraps a bidirectional streaming function in order to record trace spans.
+ * @param {cls.Namespace} namespace The CLS namespace.
+ * @param {Object} handlerSet An object containing references to the function handle,
+ * as well as serialize and deserialize handles.
+ * @param {string} requestName The human-friendly name of the request.
+ */
+function wrapBidi(namespace, handlerSet, requestName) {
+  var rootContext;
+  // CLS context object
+  var context = {};
```

Comment:
Could we add a brief comment here on why we need to explicitly chain contexts using bind instead of the run pattern used in the other hooks?